Integrate the `unyt` package

rmatsum836 opened the pull request:
This is my second attempt at integrating the Unyt package into mosdef_cassandra.  This time, I'm trying to be much more general in method of checking that the user arguments to `moves` and `system` are valid `unyt_arrays`.  

Unyt already has good features for these unit checks: https://unyt.readthedocs.io/en/stable/modules/unyt.dimensions.html#unyt.dimensions.accepts.  The `@accepts` decorator takes in a `unyt.dimension` as an argument and will check that the `unyt_array` passed in has valid units.

So far, I've written a general function `validate_unit` that does 2 checks:

1. Checks that the argument passed in is a `unyt_array`
2. Uses the `@accepts` decorator to ensure that the `unyt_array` is a valid unit

I've also started implementing `validate_unit` into `moves.py` so you can an example.  Let me know if you have any thoughts or comments on this implementation.

rmatsum836 commented:
Checklist:

- [ ] Fully remove `NM_to_A` conversion in `inp_functions`
- [ ] Ensure all arguments with units are checked (restricted_insertions)
- [ ] Update examples with unyt
- [ ] Write unit tests to make sure conversions are correct

rsdefever commented on mosdef_cassandra/writers/inp_functions.py:
```diff
@@ -236,37 +240,44 @@ def generate_input(system, moves, run_type, run_length, temperature, **kwargs):
     boxes = []
     for box in system.boxes:
         if isinstance(box, mbuild.Compound):
-            box_dims = np.hstack((box.periodicity, box.boundingbox.angles))
```
What am I missing with the change here?

rmatsum836 replied:
This was something I initially changed and need to change back.

rsdefever commented on mosdef_cassandra/writers/inp_functions.py:
```diff
@@ -1899,3 +1922,46 @@ def _check_restricted_insertions(box, restriction_type, restriction_value):
+    """Convert kwargs that are unyt units
+    """
+    if "vdw_cutoff" in kwargs:
+        kwargs["vdw_cutoff"] = kwargs["vdw_cutoff"].to("angstrom")
+    if "vdw_cutoff_box1" in kwargs:
+        kwargs["vdw_cutoff_box1"] = kwargs["vdw_cutoff_box2"].to("angstrom")
+    if "vdw_cutoff_box2" in kwargs:
+        kwargs["vdw_cutoff_box2"] = kwargs["vdw_cutoff_box2"].to("angstrom")
+    if "charge_cutoff" in kwargs:
+        kwargs["charge_cutoff"] = kwargs["charge_cutoff"].to("angstrom")
+    if "rcut_min" in kwargs:
+        kwargs["rcut_min"] = kwargs["rcut_min"].to("angstrom")
+    if "pressure" in kwargs:
+        kwargs["pressure"] = kwargs["pressure"].to("bar")
+    if "pressure_box1" in kwargs:
+        kwargs["pressure_box1"] = kwargs["pressure_box1"].to("bar")
+    if "pressure_box2" in kwargs:
+        kwargs["pressure_box2"] = kwargs["pressure_box2"].to("bar")
+    if "chemical_potentials" in kwargs:
+        new_mu = list()
+        for mu in kwargs["chemical_potentials"]:
+            mu = mu.to("kJ/mol")
+            new_mu.append(mu)
+        kwargs["chemical_potentials"] = new_mu
+    if "cbmc_rcut" in kwargs:
+        kwargs["cbmc_rcut"] = kwargs["cbmc_rcut"].to("angstrom")
+
+    return kwargs
+
+
+def _convert_moves_units(moves):
+    new_max_volume = list()
+    for max_vol in moves.max_volume:
+        max_vol = max_vol.to("angstrom**3")
+        new_max_volume.append(max_vol)
+    moves.max_volume = new_max_volume
+
+    # Add conversion for restricted_value
+
+    return moves
```
Should these functions go in their own file? If someone wants to see the unit conversions we are doing I think it would be beneficial for them to be easy to find.

rmatsum836 replied:
I could move these to `utils/unit.py`, which might be easier to find.

rmatsum836 commented:
>It seems a bit odd to have validate_unit(temperature) called right before validate_kwarg_units. Could those be combined in a clean way?

I can look into.  The reason I did it this way is that temperature is a required argument and considered to to be separate from the kwargs.  But I can try to combine these checks into a single function if you would prefer that way.

>Is there anything specific you have questions about? If not I'd say you're on the right track.

I don't think so currently.  I'm more familiar with the package now so I have a better idea of how everything works.  I'll let you know though if I run into any issues.

rmatsum836 commented:
I had a question regarding several of the tests in `test_writers.py` that are meant to trigger an error because the wrong type was passed.  For example in `test_pressure`, the pressure argument is set as `"string"`.  I have a check to make sure the pressure argument is the correct type, however this is in `runners.py`.  Based on my understanding, a user won't call `generate_input` directly.  As a result, is it okay if these checks get moved to `test_runners`? 

The other alternative is to move all the checks to the `inp_functions.py` module.

rmatsum836 commented:
Tests are passing locally for me now.  The bleeding edge checks on AZP are failing because unyt isn't being installed for some reason.  I'l figure this out later.

rsdefever commented:
@rmatsum836 it looks like in the azure-pipelines.yml that the `mosdef_cassandra` requirements.txt are not installed. Adding that line should fix the problem.

rsdefever commented on mosdef_cassandra/writers/inp_functions.py:
```diff
@@ -31,7 +31,7 @@ def generate_input(
         simulation length in units of (default=steps), or your choice
         as specified by the 'units' option in **kwargs
-    temperature : unyt array
+    temperature : unyt_array
```
Would it be more clear to call these `unyt_quantity`; I know a `unyt_quantity` is a `unyt array`. Just wondering if the terminology is confusing.

rmatsum836 replied:
I changed these to say `unyt_array or unyt_quantity` just to be clear.  From my understanding, these are pretty interchangeable.  For example, `isinstance(1 * u.nm, u.unyt_array` and `isinstance(1 * u.nm, u.array.unyt_quantity)` both return True

rmatsum836 commented:
Good catch on `max_dihedral`, I added the following:
- `max_dihedral` is now required to be a unyt quantity or array
- Check existed to ensure that max_dihedral was greater than 0, added a check to ensure max_dihedral is less than 360

I also added several unit tests to check the `validate_unit` function.
